README.md
#東京電力 Watch

東京電力管内の電力状況に関する以下の情報をポップアップ表示する Google Chrome 用のブラウザ拡張です。

- 現在および過去の電力使用量
- 今後予測される電力使用量
- 指定したグループの計画停電の実施状況および予定
- 最新の停電情報や節電情報に関する Twitter 上のツイート

電力使用状況のデータ取得には [東京電力電力供給状況API](http://tepco-usage-api.appspot.com/) を利用しています。
昨日以前のデータを取得してグラフ表示することもできます。

今後の電力使用量予報の取得には [Yahoo! 電気予報API](http://developer.yahoo.co.jp/webapi/shinsai/v1/setsuden/electricpowerforecast.html) を利用しています。

計画停電情報の取得には [計画停電情報API](http://mukku.org/v2.00/) を利用しています。

また [@OfficialTEPCO](https://twitter.com/OfficialTEPCO) と [@Yahoo_DenkiYoho](https://twitter.com/Yahoo_DenkiYoho) のツイートを取得して、最新の停電情報や節電情報を表示します。
ツイートを取得するリストを自由に変更することもできます。

電力使用量やツイートの情報が更新されたときは、デスクトップ通知でお知らせします。
電力使用量が一定値以上のときにデスクトップ通知するなどの設定ができます。

グラフ表示には [jqPlot](http://www.jqplot.com/) 1.0.4 を利用しています。

Windows 7, Mac OS X で動作確認しています。
不具合報告などは [@tichi73](http://twitter.com/tichi73) までお願いします。

ChangeLog.txt
tepco_watch

v0.7.0: YYYY/MM/DD
- Chrome extension manifest version2 に対応しました。
- jqPlot を version 1.0.4 に更新しました。

v0.6.1: 2011/07/04
- 使用率予報が正しく表示できないことがある不具合を修正しました。
- 東京電力による使用率予報が提供されている場合は、Yahoo! による予報より優先して表示するようにしました。
- バッジアイコン上での電力使用率の表示方法を変更しました。

v0.6.0: 2011/07/01
- 一番上に表示される電力使用量に、5分ごとに更新される速報値を使うようにしました。
- 計画停電情報APIを利用して、計画停電が予定されているか表示するようにしました。

v0.5.2: 2011/05/22
- オプション画面の数値入力ボックスに min/max 属性を設定しました。

v0.5.3: 2011/06/21
- proxy内部からtwitterにアクセスできないことがあり、アクセス方法を変更しました。


options.js
(function(){
var option_notify = {
	title: "デスクトップ通知に関する設定",
	options: [
		{
			id: "notify_enabled",
			legend: "新しい情報を取得したときのデスクトップ通知",
			type: "select",
			select: [
				{text: "有効", value: true},
				{text: "無効", value: false}
			],
			class: "select select_boolean",
		},{
			id: "notify_anytime",
			legend: "常に通知を行う（注意：テスト用）",
			type: "select",
			select: [
				{text: "通知する", value: true},
				{text: "通知しない", value: false}
			],
			class: "select select_boolean",
			enabled_if: function(){ return $("#notify_enabled").val()=="true" },
		},{
			id: "notify_timeout",
			legend: "デスクトップ通知を自動的に閉じる時間",
			type: "number",
			min: 3,
			max: 600,
			size: 3,
			postfix: "秒",
			class: "input_seconds",
			enabled_if: function(){ return $("#notify_enabled").val()=="true" },
		},
	],
};
var option_usage = {
	title: "<a href=\"http://tepco-usage-api.appspot.com/\" target=\"_blank\">東京電力電力供給状況API</a> による情報取得に関する設定",
	options: [
		{
			id: "latest_update_interval",
			legend: "電力使用状況を取得する間隔",
			type: "number",
			min: 5,
			max: 120,
			size: 3,
			postfix: "分",
			class: "input_minutes",
		},{
			id: "notify_threshold",
			legend: "デスクトップ通知を行う電気使用量しきい値",
			type: "number",
			min: 0,
			max: 100,
			size: 3,
			postfix: "％以上のとき通知する",
			class: "input_number",
		},{
			id: "notify_in_saving",
			legend: "計画停電実施中のときのデスクトップ通知",
			type: "select",
			select: [
				{text: "通知する", value: true},
				{text: "通知しない", value: false}
			],
			class: "select select_boolean",
			enabled_if: function(){ return $("#notify_enabled").val()=="true" },
		},
	],
};
var option_forecast = {
	title: "<a href=\"http://developer.yahoo.co.jp/webapi/shinsai/v1/setsuden/electricpowerforecast.html\" target=\"_blank\">Yahoo! 電気予報API</a> による情報取得に関する設定",
	options: [
		{
			id: "forecast_enabled",
			legend: "電気予報APIによる情報の取得",
			type: "select",
			select: [
				{text:"取得する", value: true},
				{text:"取得しない", value: false}
			],
			class: "select select_boolean",
		},{
			id: "forecast_update_interval",
			legend: "電気予報を取得する間隔",
			type: "number",
			min: 1,
			max: 24,
			size: 3,
			postfix: "時間",
			class: "input_hours",
			enabled_if: function(){ return $("#forecast_enabled").val()=="true" },
		},
	],
}
var option_twitter = {
	title: "Twitter による情報取得に関する設定",
	options: [
		{
			id: "twitter_enabled",
			legend: "ツイッターから関連ツイートを以下リストから取得",
			type: "select",
			select: [
				{text:"取得する", value: true},
				{text:"取得しない", value: false}
			],
			class: "select select_boolean",
		},{
			id: "twitter_username",
			legend: "＠ユーザ名／リスト名",
			type: "text",
			size: 30,
			prefix: '@',
			postfix: ' <span id="check_twitter">確認する</span>',
			enabled_if: function(){ return $("#twitter_enabled").val()=="true" },
		},{
			id: "twitter_update_interval",
			legend: "ツイートを取得する間隔",
			type: "number",
			min: 5,
			max: 120,
			size: 3,
			postfix: "分",
			class: "input_minutes",
			enabled_if: function(){ return $("#twitter_enabled").val()=="true" },
		},{
			id: "twitter_notify",
			legend: "新しいツイートがあったときのデスクトップ通知",
			type: "select",
			select: [
				{text:"通知する", value: true},
				{text:"通知しない", value: false}
			],
			class: "select select_boolean",
			enabled_if: function(){ return $("#notify_enabled").val()=="true" && $("#twitter_enabled").val()=="true" },
		},{
			id: "twitter_view_min",
			legend: "少なくとも表示するツイートの個数",
			type: "number",
			min: 1,
			max: 100,
			size: 3,
			postfix: "ツイート",
			class: "input_number",
			enabled_if: function(){ return $("#twitter_enabled").val()=="true" },
		},{
			id: "twitter_view_duration",
			legend: "これより古いツイートは表示しない",
			type: "number",
			min: 1,
			max: 720,
			size: 3,
			postfix: "時間",
			class: "input_hours",
			enabled_if: function(){ return $("#twitter_enabled").val()=="true" },
		},
	],
};

var option_graph = {
	title: "電力使用状況のグラフ表示に関する設定",
	options: [
		{
			id: "graph_enabled",
			legend: "電力使用状況のグラフ表示",
			type: "select",
			select: [
				{text:"表示する", value: true},
				{text:"表示しない", value: false}
			],
			class: "select select_boolean",
		},{
			id: "graph_fx_off",
			legend: "グラフ切り替え時のアニメーション効果",
			type: "select",
			select: [
				{text:"効果あり", value: false},
				{text:"効果なし", value: true}
			],
			class: "select select_boolean",
			enabled_if: function(){ return $("#graph_enabled").val()=="true" },
		},
	],
};

var option_mukku = {
	title: "計画停電情報APIに関する設定",
	title: "<a href=\"http://mukku.org/v2.00/\" target=\"_blank\">計画停電情報API</a> による情報取得に関する設定",
	options: [
		{
			id: "mukku_enabled",
			legend: "計画停電情報APIによる情報の取得",
			type: "select",
			select: [
				{text:"取得する", value: true},
				{text:"取得しない", value: false}
			],
			class: "select select_boolean",
		},{
			id: "mukku_update_interval",
			legend: "計画停電情報を取得する間隔",
			type: "number",
			min: 1,
			max: 24,
			size: 3,
			postfix: "時間",
			class: "input_hours",
			enabled_if: function(){ return $("#mukku_enabled").val()=="true" },
		},{
			id: "mukku_group",
			legend: "計画停電のグループ番号（<a href='http://www.tepco.co.jp/KT/' target='_blank'>東京電力HPで調べる</a>）",
			type: "select",
			select: [
				{text:"（未設定）", value: ""},
				{
					optgroup: "1グループ",
					select: [
						{text:"1-A", value: "1-A"},
						{text:"1-B", value: "1-B"},
						{text:"1-C", value: "1-C"},
						{text:"1-D", value: "1-D"},
						{text:"1-E", value: "1-E"},
					]
				},
				{
					optgroup: "2グループ",
					select: [
						{text:"2-A", value: "2-A"},
						{text:"2-B", value: "2-B"},
						{text:"2-C", value: "2-C"},
						{text:"2-D", value: "2-D"},
						{text:"2-E", value: "2-E"},
					]
				},
				{
					optgroup: "3グループ",
					select: [
						{text:"3-A", value: "3-A"},
						{text:"3-B", value: "3-B"},
						{text:"3-C", value: "3-C"},
						{text:"3-D", value: "3-D"},
						{text:"3-E", value: "3-E"},
					]
				},
				{
					optgroup: "4グループ",
					select: [
						{text:"4-A", value: "4-A"},
						{text:"4-B", value: "4-B"},
						{text:"4-C", value: "4-C"},
						{text:"4-D", value: "4-D"},
						{text:"4-E", value: "4-E"},
					]
				},
				{
					optgroup: "5グループ",
					select: [
						{text:"5-A", value: "5-A"},
						{text:"5-B", value: "5-B"},
						{text:"5-C", value: "5-C"},
						{text:"5-D", value: "5-D"},
						{text:"5-E", value: "5-E"},
					]
				},
			],
			class: "select",
			enabled_if: function(){ return $("#mukku_enabled").val()=="true" },
		},
	],
};

var option_list = [
	option_notify,
	option_usage,
	option_forecast,
	option_twitter,
	option_graph,
	option_mukku,
];

function create_option_table(opt) {
	var templete_tbody = $("#templete_tbody");
	var option_tbody = $(".option_tbody");

	var tr = $(".title_tr", templete_tbody).clone(true);
	$(".option_title", tr).append(opt.title);
	option_tbody.append(tr);

	for (var i=0 ; i<opt.options.length ; i++) {
		var o = opt.options[i];
		var tr = $(".option_tr", templete_tbody).clone(true);
		$(".option_desc", tr).append(o.legend);
		var input;
		switch (o.type) {
		case "number":
		case "text":
			input = $(document.createElement('input'));
			input.attr("type", o.type).attr("id", o.id);
			if (o.size) {
				input.attr("size", o.size);
			}
			if (o.maxlength) {
				input.attr("maxlength", o.size);
			}
			if (o.type == "number") {
				if ("min" in o) {
					input.attr("min", o.min);
				}
				if ("max" in o) {
					input.attr("max", o.max);
				}
			}
			break;
		case "select":
			input = $(document.createElement('select'));
			input.attr("id", o.id);
			for (var j=0 ; j<o.select.length ; j++) {
				if (o.select[j].optgroup) {
					var optgroup = $(document.createElement('optgroup'));
					var select = o.select[j].select;
					optgroup.attr('label', o.select[j].optgroup);
					for (var k=0 ; k<select.length ; k++) {
						var option = $(document.createElement('option'));
						option.val(select[k].value).text(select[k].text);
						optgroup.append(option);
					}
					input.append(optgroup);
				} else {
					var option = $(document.createElement('option'));
					option.val(o.select[j].value).text(o.select[j].text);
					input.append(option);
				}
			}
			break;
		}
		if (input) {
			if (o.class) {
				input.addClass(o.class);
			}
			if (o.css) {
				input.css(o.css);
			}
			$(".option_input", tr).append(input);
			inputs[o.id] = input;
		}
		if (o.prefix) {
			$(".option_prefix", tr).append(o.prefix);
		}
		if (o.postfix) {
			$(".option_postfix", tr).append(o.postfix);
		}
		if ("min" in o || "max" in o) {
			var range = "（";
			if ("min" in o) {
				range += o.min;
			}
			range += "～";
			if ("max" in o) {
				range += o.max;
			}
			range += "）";
			$(".option_range", tr).text(range);
		}
		option_tbody.append(tr);
	}
}

function check_option_enabled(opt) {
	for (var i=0 ; i<opt.options.length ; i++) {
		var o = opt.options[i];
		var input = $("#" + o.id);
		if (!input) {
			continue;
		}
		if (o.type == "number") {
			var value = parseInt(input.val());
			// console.log(value);
			if (isNaN(value)) {
				value = config.__defmap[o.id];
				if (input.hasClass('input_seconds')) {
					value /= 1000;
				} else if (input.hasClass('input_minutes')) {
					value /= 60 * 1000;
				} else if (input.hasClass('input_hours')) {
					value /= 60 * 60 * 1000;
				}
				// console.log(o.id + "(def)=" + value);
			}
			if ('min' in o && value < o.min) {
				value = o.min;
			}
			if ('max' in o && value > o.max) {
				value = o.max;
			}
			input.val(value);
		}
		if (o.enabled_if) {
			var enabled = o.enabled_if();
			input.prop("disabled", !enabled);
		}
	}
}

function create_option() {
	for (var i=0 ; i<option_list.length ; i++) {
		create_option_table(option_list[i]);
	}
}

function check_option() {
	for (var i=0 ; i<option_list.length ; i++) {
		check_option_enabled(option_list[i]);
	}
}

function onchange() {
	// console.log("========== onchange ==========");
	var input = $(this);
	var id = input.attr('id');
	var value = input.val();
	// console.log(id + "=" + value);
	input.addClass('changed');
	if (id == 'twitter_username') {
		value = value.replace(/[^\w\-\/]/g,'');
		value = value.replace(/\/$/,'');
		input.val(value);
	}
	check_option();
}

function check_twitter() {
	var value = $('#twitter_username').val();
	var url = 'https://twitter.com/';
	if (value.match(/(.+)\/(.+)/)) {
		url += '#!/list/' + value;
	} else {
		url += '#!/' + value;
	}
	window.open(url, 'check_twitter', 'width=1024, height=768');
}

function load_option() {
	// console.log("========== load_option ==========");
	config.load();
	setup_option();
}

function reset_option() {
	// console.log("========== reset_option ==========");
	config.resetToDefault();
	setup_option();
}

function setup_option() {
	// console.log("========== setup_option ==========");
	for (var id in inputs) {
		var input = inputs[id];
		var value = config[id];
		// console.log(id + "=" + value);
		if (input.hasClass('input_seconds')) {
			value /= 1000;
		} else if (input.hasClass('input_minutes')) {
			value /= 60 * 1000;
		} else if (input.hasClass('input_hours')) {
			value /= 60 * 60 * 1000;
		}
		input.val(value.toString());
		input.removeClass('changed');
	}
	check_option();
}

function save_option() {
	// console.log("========== save_option ==========");
	for (var id in inputs) {
		var input = inputs[id];
		var value = input.val();
		if (input.hasClass('input_seconds')) {
			value *= 1000;
		} else if (input.hasClass('input_minutes')) {
			value *= 60 * 1000;
		} else if (input.hasClass('input_hours')) {
			value *= 60 * 60 * 1000;
		} else if (input.hasClass('input_number')) {
			value = parseInt(value);
		} else if (input.hasClass('select_boolean')) {
			value = value != 'false';
		}
		// console.log(id + "=" + value);
		input.removeClass('changed');
		config[id] = value;
	}
	config.save();
	var options_port = chrome.extension.connect({name: 'options'});
	options_port.postMessage({action: "reqConfigReload"});
	options_port.disconnect();
}

var config = new TepcoWatcherConfig;
var inputs = {};

$(document).ready(function(){
	create_option();
	load_option();
	$('#load').click(load_option);
	$('#save').click(save_option);
	$('#reset').click(reset_option);
	$('input').change(onchange);
	$('select').change(onchange);
	$('#check_twitter').click(check_twitter);
});
})();


popup.js
(function(){
//---------------------------------------------------------
// Status
//---------------------------------------------------------
PopupStatus = function() {
	this.define('compare_yesterday'		,true);
	this.define('compare_lastweek'		,false);
	this.define('percent_mode'		,false);
	this.define('bottomcut_mode'		,false);
	this.__name = 'popup';
	this.__autosave = true;
}
PopupStatus.prototype = new Config();

//---------------------------------------------------------
var max_capacity_default = 6000;
var y_tickInterval = 1000;
var busy = { latest: false, usage: false };

var view_date;
var move_date;
var graph_side = 0;
var tweet_pasttime_timer;
var preloading = true;

var config = new TepcoWatcherConfig;
var status = new PopupStatus;

var popup_port;

const DATE_1DAY_OFFSET	= 24 * 60 * 60 * 1000;

const KEYCODE_LEFT	= 37;
const KEYCODE_UP	= 38;
const KEYCODE_RIGHT	= 39;
const KEYCODE_DOWN	= 40;

function jqplot_today(arg) {
	var config = {
		label: '当日実績', shadow: false, renderer: $.jqplot.BarRenderer, rendererOptions: {barMargin: 14}
	};
	for (var i in arg) { config[i] = arg[i]; }
	return config;
}

function jqplot_yesterday(arg) {
	var config = {
		label: '前日実績', markerOptions: {size: 6}
	};
	for (var i in arg) { config[i] = arg[i]; }
	return config;
}

function jqplot_capacity(arg) {
	var config = {
		label: '供給力', showLabel: false, lineWidth: 1, markerOptions: {show: true, size: 3}
	};
	for (var i in arg) { config[i] = arg[i]; }
	return config;
};

function jqplot_default(arg) {
	var config = {
		seriesColors: [ "#FF0000", "#4bb2c5", "#c5b47f", "#EAA228", "#579575", "#839557", "#958c12", "#953579", "#4b5de4", "#d8b83f", "#ff5800", "#0085cc"],
		legend: {
			show: true, location: 'sw',
			// xoffset: 6, yoffset: 6,
		},
		series: [],
		axes: {
			xaxis: {
				renderer: $.jqplot.CategoryAxisRenderer, 
				ticks: (function(){var a=[];for(var i=0;i<24;i++){a.push(i.toString())};return a})(),
			},
			yaxis: {
				min: 0,
				max: max_capacity_default,
				tickOptions: { formatString: '%d' },
				tickInterval: y_tickInterval,
			},
		},
		cursor: {  
			show: true,
			showVerticalLine: false,
			showHorizontalLine: true,
			showCursorLegend: false,
			showTooltip: false,
			zoom: true,
			dblClickReset: true,
		},
	};
	for (var i in arg) { config[i] = arg[i]; }
	return config;
};

function setBusy(type) {
	if (busy[type]) {
		return -1;
	}
	busy[type] = true;
	switch (type) {
	case 'usage':
		// $('.navi_button').attr("disabled","disabled");
		break;
	}
	return 0;
}

function clearBusy(type) {
	busy[type] = false;
	switch (type) {
	case 'usage':
		// $('.navi_button').removeAttr("disabled");
		break;
	}
}

function sendMessage(msg) {
	if (popup_port) {
		popup_port.postMessage(msg);
	} else {
		console.error("popup_port not connected!");
	}
}

function onMessage(request, sender, callback) {
	switch (request.action) {
	case 'rspQuickData':
		updateQuickView(request.data, function(){ clearBusy('quick'); });
		break;

	case 'rspLatestData':
		updateLatestView(request.data, function(){ clearBusy('latest'); resize(); });
		break;

	case 'rspUsageData':
		updateGraphView(request.data, function(){ clearBusy('usage') });
		break;

	case 'rspTweetData':
		updateTweetView(request.data, function(){ clearBusy('tweet') });
		break;

	case 'rspMukkuData':
		updateMukkuView(request.data, function(){ clearBusy('mukku') });
		break;
	}
}

function updateQuickView(data, callback) {
	// 使用率速報値表示
	var quick_usage = data.usage + ' 万kW【' + data.usage_rate.toFixed(1) + '%】使用中 [' + data.date + ']';
	$('#usage').text(quick_usage);
}

function updateLatestView(data, callback) {
	// アイコンの設定
	if (data.saving) {
		$('#icon').prop("src", "img/icon_off48.png");
		$('#saving').prop('class', 'saving_on').text("計画停電：実施中です！");
	} else {
		$('#icon').prop("src", "img/icon_on48.png");
		$('#saving').prop('class', 'saving_off').text("計画停電：実施していません");
	}

	// 予想最大電力／ピーク時供給力
	var capacity_info = '予想需要 ' + data.forecast_peak_usage + ' 万kW(' + data.forecast_peak_period + '時台)';
	capacity_info += ' ／ ';
	capacity_info += '最大供給力 ' + data.capacity + ' 万kW(' + data.capacity_peak_period + '時台)';
	$('#capacity_info').text(capacity_info);

	// データ読み込み日時の更新
	var usage_updated = new Date(data.usage_updated);
	$("#updated").text(usage_updated.toLocaleDateString() + ' ' + usage_updated.toLocaleTimeString());

	// グラフ表示日が未設定の場合、最新日に移動（初回時・更新時）
	if (!view_date) {
		view_date = new Date(data.year, data.month-1, data.day);
		if (config.graph_enabled) {
			moveDate(0);
		} else {
			preloadingDone();
		}
	}
	callback();
}

function preloadingNone() {
	if (preloading) {
		$('#preloading_div').hide();
		preloading = false;
	}
}
function preloadingDone() {
	if (preloading) {
		if (config.graph_fx_off) {
			$('#preloading_div').hide();
		} else {
			$('#preloading_div').fadeOut(config.graph_fadetime);
		}
		preloading = false;
	}
}

function updateGraphView(data, callback) {
	if (!config.graph_enabled) {
		return;
	}
	var y_max = 0;
	var y_min = 10000000;
	var y_data = [];
	var plot_date;
	var plot_capacity = status.percent_mode ? false : true;
	var jqplot_config = jqplot_default();
	
	function check_max(y) {
		if (y_max < y) {
			y_max = y;
		}
		if (y_min > y) {
			y_min = y;
		}
	}
	function date_label(date, cnt0, cnt1, cnt2, cnt3) {
		var label = date.toLocaleDateString();
		var opt = '';
		if (cnt0 > 0) {
			opt += '【停電】';
		}
		if (cnt1 > 0) {
			opt += '［実績］';
		}
		if (cnt2 > 0) {
			opt += '［でんき予報］';
		}
		if (cnt3 > 0) {
			opt += '［Yahoo!予報］';
		}
		if (opt) {
			label += ' ' + opt;
		}
		return label;
	}
	function graph_toggleside() {
		graph_side = graph_getside(false);
	}
	function graph_getside(currentside) {
		return currentside ? graph_side : ((graph_side + 1) & 1);
	}
	function graph_getid(currentside) {
		var side = graph_getside(currentside);
		return 'graph' + side;
	}
	function data_prepare() {
		// ロードしたデータを jqPlot に渡す準備を行う
		for (var i=0 ; i<data.length ; i++) {
			var usage = data[i];
			if (!usage || usage.length == 0) {
				continue;
			}
			var date = new Date(usage[0].year, usage[0].month-1, usage[0].day);
			if (i == 0) {
				// 当日データの場合（配列の先頭を当日としている）
				var y_usages = [[],[],[],[]];
				var y_count = [0,0,0,0];
				var y_capacity = [];
				for (var t=0 ; t<usage.length ; t++) {
					var y_value0 = usage[t].usage || usage[t].forecast;
					var y_value = status.percent_mode ? (100 * y_value0 / usage[t].capacity) : y_value0;
					var type = usage[t].usage_type;
					y_usages[0].push(type == 0 ? y_value : 0);
					y_usages[1].push(type == 1 ? y_value : 0);
					y_usages[2].push(type == 2 ? y_value : 0);
					y_usages[3].push(type == 3 ? y_value : 0);
					++y_count[type];
					if (plot_capacity) {
						y_capacity.push(usage[t].capacity);
						check_max(usage[t].capacity);
					}
					check_max(y_value0);
				}

				jqplot_config.series.push(jqplot_today(y_count[0] ? {label: date_label(date,1,0,0,0)}  : {showLabel: false} ));
				jqplot_config.series.push(jqplot_today(y_count[1] ? {label: date_label(date,0,1,0,0)}  : {showLabel: false} ));
				jqplot_config.series.push(jqplot_today(y_count[2] ? {label: date_label(date,0,0,1,0)}  : {showLabel: false} ));
				jqplot_config.series.push(jqplot_today(y_count[3] ? {label: date_label(date,0,0,0,1)}  : {showLabel: false} ));

				y_data.push(y_usages[0]);
				y_data.push(y_usages[1]);
				y_data.push(y_usages[2]);
				y_data.push(y_usages[3]);

				if (plot_capacity) {
					jqplot_config.series.push(jqplot_capacity());
					y_data.push(y_capacity);
				}
				plot_date = date;
			} else {
				var y_usage = [];
				var y_count = [0,0,0,0];
				for (var t=0 ; t<usage.length ; t++) {
					var y_value0 = usage[t].usage || usage[t].forecast;
					var y_value = status.percent_mode ? (100 * y_value0 / usage[t].capacity) : y_value0;
					y_usage.push(y_value);
					check_max(y_value);
					++y_count[usage[t].usage_type];
				}
				var series = jqplot_yesterday({label: date_label(date, y_count[0], y_count[1], y_count[2], y_count[3])});
				jqplot_config.series.push(series);
				y_data.push(y_usage);
			}
		}
	}

	function draw_graph () {
		if (plot_date && y_data.length > 0) {
			// jqplot_config.title = view_date.toLocaleDateString(true);
			// Y軸最大値を調整
			if (status.percent_mode) {
				jqplot_config.axes.yaxis.max = 100;
				jqplot_config.axes.yaxis.tickInterval = 10;
				if (status.bottomcut_mode) {
					jqplot_config.axes.yaxis.min = 50;
					jqplot_config.axes.yaxis.tickInterval = 5;
				}
			} else {
				// jqplot_config.axes.yaxis.max = y_max + (y_tickInterval / 2);
				// jqplot_config.axes.yaxis.max = y_max;
				var y_tick = y_tickInterval;
				if (status.bottomcut_mode) {
					y_tick = Math.floor(y_tick / 2);
					jqplot_config.axes.yaxis.tickInterval = y_tick;
					jqplot_config.axes.yaxis.min = Math.floor(y_min / y_tick) * y_tick;
				}
				jqplot_config.axes.yaxis.max = Math.floor((y_max + y_tick - 1) / y_tick) * y_tick;
			}
			// グラフを生成する
			new_graph.empty();
			new_graph_outer.show();	// 一旦 show しないと jqplot がグラフを描画してくれない
			$.jqplot(new_graph_id, y_data, jqplot_config);
			new_graph_outer.hide();
			// 凡例の行間を詰める
			$("td.jqplot-table-legend").css('padding-top', '2px');
			$(".jqplot-table-legend").hide();
		} else {
			var msg = "＜データがありません＞";
			var table = '<table width="100%" height="100%" class="glaph_error">';
			table += '<tr><td align="center" valign="middle">' + msg + '</td></tr></table>';
			new_graph.empty();
			new_graph.removeAttr('class');
			new_graph.append(table);
		}
		new_graph_outer.css('z-index', 2);
		old_graph_outer.css('z-index', 1);
	}

	function draw_after() {
		view_date = move_date;
		$('#update_data').val(view_date.toLocaleDateString(true));
		old_graph_outer.hide();
		old_graph.empty();
		graph_toggleside();
 		$(".jqplot-event-canvas").hover(
			function(){ $(".jqplot-table-legend").fadeIn(); },
			function(){ $(".jqplot-table-legend").fadeOut(); }
		);
		callback();
	}

	var new_graph_id = graph_getid(true);
	var new_graph = $('#' + new_graph_id);
	var new_graph_outer = $('#' + new_graph_id + '_outer');

	var old_graph_id = graph_getid(false);
	var old_graph = $('#' + old_graph_id);
	var old_graph_outer = $('#' + old_graph_id + '_outer');

	data_prepare();
	draw_graph();

	if (preloading) {
		new_graph_outer.show();
		draw_after();
		preloadingDone();
	} else {
		if (config.graph_fx_off) {
			new_graph_outer.show();
			draw_after();
		} else {
			new_graph_outer.fadeIn(config.graph_fadetime, draw_after);
		}
	}
}

function updateMukkuView(data, callback) {
	var mi = $('#mukku_info');
	var mukku_on = false;
	var text = '';

	if (config.mukku_group == "") {
		var options_url = chrome.extension.getURL("options.html");
		var options_a = $(document.createElement('a'));
		options_a.attr('href', options_url).attr('target', '_blank').text('[オプション]');
		mi.empty();
		mi.append(options_a);
		mi.append(' から計画停電のグループ番号を設定してください');
	} else {
		var result = data.Result[config.mukku_group];
		if (!result) {
			mukku_on = true;
			text = "計画停電の情報が取得できませんでした";
		} else {
			if (result.Count > 0) {
				text = result.All + ' に計画停電が予定されています [' + config.mukku_group + ']';
				mukku_on = true;
			} else {
				text = '本日 グループ ' + config.mukku_group + ' に計画停電の予定はありません';
			}
		}
		mi.text(text);
	}
	if (mukku_on) {
		mi.addClass('mukku_on');
		mi.removeClass('mukku_off');
	} else {
		mi.removeClass('mukku_on');
		mi.addClass('mukku_off');
	}
	callback();
}

function updateTweetTime() {
	if (!config.twitter_enabled) {
		return;
	}
	function pasttimeText(msec) {
		var sec = Math.floor(msec / 1000);
		if (sec < 60) {
			return sec + '秒前';
		}
		var min = Math.floor(sec / 60); sec %= 60;
		if (min < 60) {
			return min + '分' + sec + '秒前';
		}
		var hour = Math.floor(min / 60); min %= 60;
		if (hour < 24) {
			return hour + '時間' + min + '分前';
		}
		var day = Math.floor(hour / 24); hour %= 24;
		return day + '日と' + hour + '時間前';
	}
	var now = new Date();
	var date = new Date();
	$('#tweet_table tbody').each(function(){
		var time = $(".tweet_time", this).text();
		date.setTime(time);
		var datetext = date.toLocaleDateString() + ' ' + date.toLocaleTimeString();
		datetext += ' (' + pasttimeText(now.getTime() - time) + ')';
		$(".tweet_date", this).text(datetext);
	});

	if (tweet_pasttime_timer != null) {
		clearTimeout(tweet_pasttime_timer);
	}
	tweet_pasttime_timer = setTimeout(updateTweetTime, 60 * 1000);
}

function updateTweetView(data, callback) {
	if (!config.twitter_enabled) {
		return;
	}
	var table = $('#tweet_table');
	table.empty();
	if (data && data.length > 0) {
		var tbody_templete = $('#tweet_templete tbody:first-child');
		var now = new Date();
		for (var i=0 ; i<data.length && i<config.twitter_view_max; i++) {
			var tweet = data[i];
			var date = new Date(tweet.created_at);
	 		var pasttime = now.getTime() - date.getTime();
			if (pasttime > config.twitter_view_duration && i >= config.twitter_view_min) {
				break;
			}
			var tbody = tbody_templete.clone(true);
			$('.tweet_icon', tbody).attr('src', tweet.image_url);
			$('.tweet_link', tbody).attr('href', tweet.tweet_link);
			$('.tweet_name', tbody).attr('href', tweet.user_link).text(tweet.screen_name);
			$('.tweet_date', tbody).text("");
			$('.tweet_time', tbody).text(date.getTime());
			$('.tweet_text', tbody).text(tweet.text);
			table.append(tbody);
		}
		updateTweetTime();
	} else {
		var msg = "＜ツイートが取得できません＞";
		var tbody = '<tbody class="glaph_error"><tr><td align="center" valign="middle">' + msg + '</td></tr></tbody>';
		table.append(tbody);
	}
	callback();
}

function moveDate(move) {
	if (!config.graph_enabled) {
		return;
	}
	if (setBusy('usage') < 0) {
		return;
	}
	function offsetDate(date, offset) {
		return new Date(date.getTime() + offset * DATE_1DAY_OFFSET);
	}
	var dates_list = [];
	dates_list.add = function(date) { this.push(date.toDateString()) };

	move_date = offsetDate(view_date, move);
	dates_list.add(move_date);
	if (status.compare_yesterday) {
		dates_list.add(offsetDate(move_date, -1));
	}
	if (status.compare_lastweek) {
		dates_list.add(offsetDate(move_date, -7));
	}
	sendMessage({action: "reqUsageData", dates: dates_list});
}

function graphOption() {
	var type = this.name;
	switch (type) {
	case 'yesterday':
	case 'lastweek':
		status['compare_' + type] = this.checked;
		break;
	case 'percent':
	case 'bottomcut':
		status[type + '_mode'] = this.checked;
		break;
	default:
		return;
	}
	moveDate(0);
}

function updateQuickData() {
	setBusy('quick');
	sendMessage({action: "reqQuickData"});
}

function updateLatestData() {
	setBusy('latest');
	view_date = null;
	sendMessage({action: "reqLatestData"});
}

function updateTweetData() {
	if (!config.twitter_enabled) {
		return;
	}
	setBusy('tweet');
	sendMessage({action: "reqTweetData"});
}

function updateMukkuData() {
	setBusy('mukku');
	sendMessage({action: "reqMukkuData"});
}

function resize() {
	if (!config.twitter_enabled) {
		$('#tweet_area').hide();
	}
	if (!config.graph_enabled) {
		$('#graph_area').hide();
		$('#navi_area').hide();
	}
	if (!config.twitter_enabled && !config.graph_enabled) {
		preloadingNone();
	}
	if (!config.mukku_enabled) {
		$('#mukku_area').hide();
	}

	var table_width = $('#main_table').outerWidth(true);
	var table_height = $('#main_table').outerHeight(true);

	// console.log("resize: " + table_width + "x" + table_height);

	$(".resize_target").width(table_width);
	$(".resize_target").height(table_height);
	
	$('html').width($('body').outerWidth(true));
	$('html').height($('body').outerHeight(true));
	$(':focus').trigger('blur');
}

function init_graph_ctrl() {
	if (!config.graph_enabled) {
		return;
	}
	
	$('#chk_yesterday').prop('checked', !!status.compare_yesterday).click(graphOption);
	$('#chk_lastweek').prop('checked', !!status.compare_lastweek).click(graphOption);
	$('#chk_percent').prop('checked', !!status.percent_mode).click(graphOption);
	$('#chk_bottomcut').prop('checked', !!status.bottomcut_mode).click(graphOption);
	
	$('#move_prev').click(function(){moveDate(-1)});
	$('#move_next').click(function(){moveDate(+1)});
	$('#update_data').click(updateLatestData).val(new Date().toLocaleDateString(true));

	var panel = $('#navi_panel');
	var option = $('#navi_option');
	panel.width(option.outerWidth());
	var offset = panel.offset();
	offset.top  = offset.top  + panel.outerHeight() - option.outerHeight() - 0;
	offset.left = offset.left + panel.outerWidth()  - option.outerWidth()  + 0;
	option.offset(offset);

	$('#navi_open').hover(
		function(){ $(this).prop('class', 'navi_open_on'); },
		function(){ $(this).prop('class', 'navi_open_off'); }
	);
	$('#navi_open').click(function(){ $('#navi_option').show(); });
	$('#navi_close').click(function(){ $('#navi_option').hide(); });

	$(window).keyup(function(e){
		switch (e.keyCode) {
		case KEYCODE_LEFT:	moveDate(-1); break;
		case KEYCODE_UP:	moveDate(-7); break;
		case KEYCODE_RIGHT:	moveDate(+1); break;
		case KEYCODE_DOWN:	moveDate(+7); break;
		}
	});
}

function start() {
	popup_port = chrome.extension.connect({name: 'popup'});
	popup_port.onMessage.addListener(onMessage);
	updateQuickData();
	updateLatestData();
	updateTweetData();
	updateMukkuData();
}

$(document).ready(function(){
	config.load();
	status.load();
	$('a').click(openTab);
	resize();
	init_graph_ctrl();
	start();
});
})();
